machine_translation/__init__.py

```diff
@@ -45,131 +46,231 @@ def main(config, tr_stream, dev_stream, use_bokeh=False):
+            logger.info("Building sampling model")
+            sampling_representation = encoder.apply(
+                sampling_input, tensor.ones(sampling_input.shape))
+            generated = decoder.generate(
+                sampling_input, sampling_representation)
+            search_model = Model(generated)
+            _, samples = VariableFilter(
+                bricks=[decoder.sequence_generator], name="outputs")(
+                    ComputationGraph(generated[1]))
+
+        # Add sampling
+        if config['hook_samples'] >= 1:
+            logger.info("Building sampler")
+            extensions.append(
+                Sampler(model=search_model, data_stream=tr_stream,
+                        hook_samples=config['hook_samples'],
+                        every_n_batches=config['sampling_freq'],
+                        src_vocab_size=config['src_vocab_size']))
+
+        # Add early stopping based on bleu
+        if config['bleu_script'] is not None:
+            logger.info("Building bleu validator")
+            extensions.append(
+                BleuValidator(sampling_input, samples=samples, config=config,
+                              model=search_model, data_stream=dev_stream,
+                              normalize=config['normalized_bleu'],
+                              every_n_batches=config['bleu_val_freq']))
+
+        # Reload model if necessary
+        if config['reload']:
+            extensions.append(LoadNMT(config['saveto']))
+
+        # Plot cost in bokeh if necessary
+        if use_bokeh and BOKEH_AVAILABLE:
+            extensions.append(
+                Plot('Cs-En', channels=[['decoder_cost_cost']],
+                     after_batch=True))
+
+        # Set up training algorithm
+        logger.info("Initializing training algorithm")
+        algorithm = GradientDescent(
+            cost=cost, parameters=cg.parameters,
+            step_rule=CompositeRule([StepClipping(config['step_clipping']),
+                                     eval(config['step_rule'])()])
+        )
+
+        # Initialize main loop
+        logger.info("Initializing main loop")
+        main_loop = MainLoop(
+            model=training_model,
+            algorithm=algorithm,
+            data_stream=tr_stream,
+            extensions=extensions
+        )
+
+        # Train!
+        main_loop.run()
+
+    elif mode == 'translate':
+
+        # Create Theano variables
+        logger.info('Creating theano variables')
+        sampling_input = tensor.lmatrix('source')
+
+        # Get test set stream
+        test_stream = get_dev_stream(
+            config['test_set'], config['src_vocab'],
+            config['src_vocab_size'], config['unk_id'])
+        ftrans = open(config['test_set'] + '.trans.out', 'w')
+
+        # Helper utilities
+        sutils = SamplingBase()
+        unk_idx = config['unk_id']
+        src_eos_idx = config['src_vocab_size'] - 1
```

Comment:
Hi @orhanf I check in ```preprocess.py``` when creating dictionary, they put ```eos_idx``` to be 0. Should this be 0 instead of ```config['src_vocab_size'] - 1```?